src/data/finite/basic.lean

```diff
@@ -0,0 +1,210 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import set_theory.cardinal.finite
+
+/-!
+# Finite types
+
+This module defines a finiteness predicate on types called `finite`.
+
+The `fintype` class also represents finiteness of a type, but a key
+difference is that a `fintype` instance represents finiteness in a
+computable way: it provides an algorithm to produce a `finset` whose
+elements enumerate the terms of the given type. A `finite` instance is
+instead a mere proposition, and as such it gets to take advantage of
+proof irrelevance.
+
+One should prefer defining `finite` instances rather than noncomputable
+`fintype` instances to preserve the property that `fintype` instances
+can be used for computation.
+
+The cardinality of a finite type `α` is given by `nat.card α`.
+
+## Main definitions
+
+* `finite α` denotes that `α` is a finite type.
+* `finite.of_fintype` creates a `finite` from a `fintype.
+* `fintype.of_finite` noncomputably creates a `fintype` from a `finite`.
+* `finite_or_infinite` is that every type is either `finite` or `infinite`.
+
+-/
+
+noncomputable theory
+open_locale classical
+
+variables {α β γ : Type*}
+
+/-- A type is `finite` if it is in bijective correspondence to some
+`fin n`.
+
+While this could be defined as `nonempty (fintype α)`, it is defined
+in this way to allow there to be `finite` instances for propositions.
+-/
+class finite (α : Sort*) : Prop :=
+(exists_equiv_fin [] : ∃ (n : ℕ), nonempty (α ≃ fin n))
```

Comment:
I wonder if this becomes easier or harder to use as
```suggestion
class inductive finite (α : Sort*)
| intro {n : ℕ} (e : α ≃ fin n) : finite α
```
You can always add `@[mk_iff]` to translate between them.

Reply:
It doesn't really change much since the main way you create a `finite` instance is `apply_instance`, but I switched it to a `class inductive` since it doesn't hurt, and there are a few places with less nesting of angle brackets.